_posts/2025-09-05-monitoring-of-custom-resources.md

```diff
@@ -0,0 +1,535 @@
+---
+layout: post
+title:  "Monitoring of custom resources"
+date: 2025-09-05
+author: sebastian_gaiser_hetzner
+---
+
+Running Kafka on Kubernetes with Strimzi is straightforward.
+But getting a complete, consistent view of everything in your Kubernetes and Kafka cluster(s) can still be hard.
+Kafka topics and users are often owned by different teams and configured differently, especially when sharing a Kafka cluster e.g. across a company.
+As usage grows, gathering information and troubleshooting gets harder.
+
+This post shows a practical, low‑friction way to close monitoring gaps using [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) so you can track every Strimzi custom resource (CR) with a simple, consistent, Kubernetes‑native approach.
+
+### What Strimzi exposes today
+
+All Strimzi operators (Cluster, User, and Topic Operator) expose metrics in Prometheus compatible format.
+However, today only the Cluster Operator exports a metric (`strimzi_resource_state`) that describes the state of custom resources (ready or not).
+
+It provides this metric for:
+
+- `Kafka`
+- `KafkaConnect`
+- `KafkaBridge`
+- `KafkaMirrorMaker2`
+- `KafkaRebalance`
+
+It does not provide this metric for:
+
+- `KafkaNodePool`
+- `KafkaConnector`
+- `StrimziPodSet`
```

Comment:
I am not sure I would mention `StrimziPodSet` here because it's an internal resource and I would not expect users to be able monitoring it. I would focus on the users facing custom resources.

Reply:
@ppatierno Why do you expect that users would not be able to monitor it? I'm not sure if StrimziPodSets give users some useful information (it is tracking some rolling update progress, but not sure it is useful on its own). But I do not see why they would not be able to monitor it.

Keep in mind it is also what is being annotated for manual rolling updates etc. So while the users are not expected and should not create or edit them (other than the annotation), it is expected that they know it exists and can monitor it.